src/bacnet/basic/object/bi.c

```diff
@@ -825,6 +858,106 @@ int Binary_Input_Read_Property(BACNET_READ_PROPERTY_DATA *rpdata)
             apdu_len =
                 encode_application_character_string(&apdu[0], &char_string);
             break;
+#if (BINARY_INPUT_INTRINSIC_REPORTING)
+        case PROP_ALARM_VALUE:
+            /* note: you need to look up the actual value */
+            apdu_len = encode_application_enumerated(
+                &apdu[0], pObject->Alarm_Value);
+            break;
+        case PROP_TIME_DELAY:
+            apdu_len =
+                encode_application_unsigned(&apdu[0], pObject->Time_Delay);
+            break;
+
+        case PROP_NOTIFICATION_CLASS:
+            apdu_len = encode_application_unsigned(
+                &apdu[0], pObject->Notification_Class);
+            break;
+
+        case PROP_EVENT_ENABLE:
+            bitstring_init(&bit_string);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_OFFNORMAL,
+                (pObject->Event_Enable & EVENT_ENABLE_TO_OFFNORMAL) ? true
+                                                                      : false);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_FAULT,
+                (pObject->Event_Enable & EVENT_ENABLE_TO_FAULT) ? true
+                                                                  : false);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_NORMAL,
+                (pObject->Event_Enable & EVENT_ENABLE_TO_NORMAL) ? true
+                                                                   : false);
+
+            apdu_len = encode_application_bitstring(&apdu[0], &bit_string);
+            break;
+
+        case PROP_EVENT_DETECTION_ENABLE:
+            apdu_len =
+                encode_application_boolean(&apdu[0], pObject->Event_Detection_Enable);
+            break;
+
+        case PROP_ACKED_TRANSITIONS:
+            bitstring_init(&bit_string);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_OFFNORMAL,
+                pObject->Acked_Transitions[TRANSITION_TO_OFFNORMAL].bIsAcked);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_FAULT,
+                pObject->Acked_Transitions[TRANSITION_TO_FAULT].bIsAcked);
+            bitstring_set_bit(&bit_string, TRANSITION_TO_NORMAL,
+                pObject->Acked_Transitions[TRANSITION_TO_NORMAL].bIsAcked);
+
+            apdu_len = encode_application_bitstring(&apdu[0], &bit_string);
+            break;
+
+        case PROP_NOTIFY_TYPE:
+            apdu_len = encode_application_enumerated(
+                &apdu[0], pObject->Notify_Type ? NOTIFY_EVENT : NOTIFY_ALARM);
+            break;
+
+        case PROP_EVENT_TIME_STAMPS:
+            /* Array element zero is the number of elements in the array */
```

Comment:
Please use the BACnetARRAY encoding helper function bacnet_array_encode() for the Event_Time_Stamps property - see example in ai.c file.
```
            apdu_len = bacnet_array_encode(
                rpdata->object_instance, rpdata->array_index,
                Analog_Input_Event_Time_Stamps_Encode,
                MAX_BACNET_EVENT_TRANSITION, apdu, apdu_size);
            if (apdu_len == BACNET_STATUS_ABORT) {
                rpdata->error_code =
                    ERROR_CODE_ABORT_SEGMENTATION_NOT_SUPPORTED;
            } else if (apdu_len == BACNET_STATUS_ERROR) {
                rpdata->error_class = ERROR_CLASS_PROPERTY;
                rpdata->error_code = ERROR_CODE_INVALID_ARRAY_INDEX;
            }
```

Reply:
Re-worked.